hail/src/main/scala/is/hail/expr/types/physical/PContainer.scala

```diff
@@ -192,7 +193,53 @@ abstract class PContainer extends PIterable {
     else
       Code(
         Region.storeInt(aoff, length),
-        Region.setMemory(aoff + const(4), nMissingBytes(length), const(if (setMissing) (-1).toByte else 0.toByte)))
+        Region.setMemory(aoff + const(4), nMissingBytes(length).toL, const(if (setMissing) (-1).toByte else 0.toByte)))
+  }
+
+  def zeroes(region: Region, length: Int): Long = {
+    val aoff = region.allocate(contentsAlignment, contentsByteSize(length))
+    initialize(region, aoff, length)
+    Region.setMemory(aoff + elementsOffset(length), length * elementByteSize, 0.toByte)
+    aoff
+  }
+
+  def zeroes(mb: MethodBuilder, region: Code[Region], length: Code[Int]): Code[Long] = {
+    val aoff = mb.newLocal[Long]
+    Code(
+      aoff := region.allocate(contentsAlignment, contentsByteSize(length)),
+      stagedInitialize(aoff, length),
+      Region.setMemory(aoff + elementsOffset(length), length.toL * elementByteSize, 0.toByte),
+      aoff)
+  }
+
+  def anyMissing(mb: MethodBuilder, aoff: Code[Long]): Code[Boolean] = {
+    if (elementType.required)
+      true
+    else {
+      val n = mb.newLocal[Int]
+      JoinPoint.CallCC[Code[Boolean]] { (jb, ret) =>
+        val loop = jb.joinPoint[Code[Int]](mb)
+        loop.define { i =>
+          (i < n).mux(
+            Region.loadByte(aoff + 4L + i.toL).cne(0).mux(
```

Comment:
An array of size 1 has 1 missing byte, with only one missing bit -- the rest of the bits are garbage memory and can be either 0 or 1, making this code correct only for sizes that are a multiple of 8.

I think it's preferable to just loop over elements and use `isElementMissing`.

Reply:
I stuck with the vectorized implementation, fixing the handling of the last chunk, and even made it iterate in 4-byte chunks. I just couldn't stomach iterating one bit at a time :smile:.